src/client/api/event.c

```diff
@@ -523,7 +523,8 @@ ev_progress_cb(void *arg)
 	/** Change status of event to INIT only if event is not in EQ and get out. */
 	if (daos_handle_is_inval(evx->evx_eqh)) {
 		D_MUTEX_LOCK(&evx->evx_lock);
-		evx->evx_status = DAOS_EVS_READY;
+		if (evx->evx_status == DAOS_EVS_COMPLETED || evx->evx_status == DAOS_EVS_ABORTED)
+			evx->evx_status = DAOS_EVS_READY;
```

Comment:
This surely seems to be an improvement; I wonder about its sufficiency though.

1. If the event is neither COMPLETED nor ABORTED after we acquire the lock, should we still return 1?
2. How can the status of the event change after the lockless checks? This makes me nervous about the lockless checks.

Reply:
1) If the status changed between the first lockless and check and this check, then it means someone had already marked the event as either complete, ready, or restarted the event and it's running. for all these cases i think returning 1 is appropriate as we don't want to continue calling progress on this event in this case.
2) it can be marked as ready by some other thread. that thread might also race ahead and run it again and make it running state. this was my main concern where this thread might acquire the lock after this and mark the event as READY while it is running. this can cause completion of the event  while the internal tse daos task is still running. finally this thread can be extremely slow and the event might have been changed back into completed state again in which case it's ready to be marked ready if this thread wakes up at that time.

we could make this safer by moving the locks to the top of the function i guess.

Reply:
it might be overkill though if we do that because those checks are just either saying continue progress (if in-flight) or stop progress (in ready state).

Reply:
OK.